### PR TITLE
This Development Cycle in Cargo 1.84

### PR DESCRIPTION
**Pull request by epage**
[Rendered](https://github.com/epage/blog.rust-lang.org/blob/cargo/posts/inside-rust/2024-12-13-this-development-cycle-in-cargo-1.84.md)

### CODE REVIEW

File: posts/inside-rust/2024-12-13-this-development-cycle-in-cargo-1.84.md [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+However, optimizations can reduce the amount of IR being processed, speeding up codegen.
+They can then also speed up proc macros, build scripts, and test runs.
+Maybe we can even design a optimization level focused on improving compile times at the cost of the debugger experience.
+Similarly, how much debug information you carry around in your binary can affect your build times.
+
+Looking at the [Rust 2023 survey results](https://blog.rust-lang.org/2024/02/19/2023-Rust-Annual-Survey-2023-results.html#challenges),
+improving compilation times and the debugging experience is neck and neck.
+The question is which debugging experience are they referring to?
+Those on the call mostly used "printf"-style debugging and would get benefit out of improving compilation time.
+Even if we surveyed people and found this representative of the Rust community (which [davidlattimore](https://github.com/davidlattimore/) did for a subset of the community on [Mastadon](https://mas.to/@davidlattimore/113484821980790635)),
```
Comment (joshtriplett) [inline posts/inside-rust/2024-12-13-this-development-cycle-in-cargo-1.84.md:257]:
```suggestion
Even if we surveyed people and found this representative of the Rust community (which [davidlattimore](https://github.com/davidlattimore/) did for a subset of the community on [Fediverse](https://mas.to/@davidlattimore/113484821980790635)),
```

(Mastodon was misspelled, but also, using the more generic term that includes other implementations.)

**Review by weihanglo — APPROVED** [review]
🎉

File: posts/inside-rust/2024-12-13-this-development-cycle-in-cargo-1.84.md [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+
+In removing the core of the custom assertions,
+we were relying on `dead_code` warnings as we removed assertions
+that were no longer used.
+However, we missed removing an assertion and epage removed it in [#14759](https://github.com/rust-lang/cargo/pull/14759).
+
+[#14781](https://github.com/rust-lang/cargo/pull/14781) and
+[#14785](https://github.com/rust-lang/cargo/pull/14785) saw us migrate the last of our "unordered lines" assertion tests.
+[#14785](https://github.com/rust-lang/cargo/pull/14785) took some investigation to figure out the best way to migrate.
+Cargo's custom assertions redacted fewwer values and allowed a test author to ignore a value redaction by using the raw value in the expected result.
```
[the remark this answers is not in the document]
Comment (weihanglo) [reply posts/inside-rust/2024-12-13-this-development-cycle-in-cargo-1.84.md:175]: I mean. Should we integrate a spell checker in this repo? 😆

File: posts/inside-rust/2024-12-13-this-development-cycle-in-cargo-1.84.md [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+
+### Rustflags and caching
+
+Cargo's [build fingerprinting](https://doc.rust-lang.org/nightly/nightly-rustc/cargo/core/compiler/fingerprint/index.html) has to satisfy several needs, including
+- Detecting when the existing build cache has to be thrown away and re-built, called the fingerprint
+- Segregating the build cache so it gets preserved across unrelated commands (e.g. alternating `cargo check` and `cargo test` without rebuilding dependencies), called [`-Cextra-filename`](https://doc.rust-lang.org/rustc/codegen-options/index.html#extra-filename)
+- Making symbol names unique so you can't unintentionally use a type in an ABI-incompatible context, called [`-Cmetadata`](https://doc.rust-lang.org/rustc/codegen-options/index.html#metadata)
+
+`RUSTFLAGS` is a way to bypass Cargo's abstractions and directly control the behavior of `rustc`.
+Cargo includes `RUSTFLAGS` in the fingerprint hash but no in the `-Cextra-filename` hash,
```
Comment (ehuss) [inline posts/inside-rust/2024-12-13-this-development-cycle-in-cargo-1.84.md:115]:
```suggestion
Cargo includes `RUSTFLAGS` in the fingerprint hash but not in the `-Cextra-filename` hash,
```

File: posts/inside-rust/2024-12-13-this-development-cycle-in-cargo-1.84.md [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+(
+[#14848](https://github.com/rust-lang/cargo/pull/14848),
+[#14846](https://github.com/rust-lang/cargo/pull/14846),
+[#14859](https://github.com/rust-lang/cargo/pull/14859)
+)
+by [epage](https://github.com/epage) and [weihanglo](https://github.com/weihanglo/),
+epage posted [#14830](https://github.com/rust-lang/cargo/pull/14830).
+However, weihanglo found there are still problems with `--remap-path-prefix`: even when using `profile.dev.split-debuginfo="packed"`, the binaries are different because the binary includes `DW_AT_GNU_dwo_name` which points to the debug file which exists per-rlib with `-Cextra-filename` included.
+
+Merging of [#14830](https://github.com/rust-lang/cargo/pull/14830) is blocked until the problem with `--remap-path-prefix` is resolved.
```
Comment (ehuss) [inline posts/inside-rust/2024-12-13-this-development-cycle-in-cargo-1.84.md:159]: I believe this was merged?

IIRC, we decided to add a hack for `--remap-path-prefix` specifically?

File: posts/inside-rust/2024-12-13-this-development-cycle-in-cargo-1.84.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (epage) [reply posts/inside-rust/2024-12-13-this-development-cycle-in-cargo-1.84.md:175]: I ran my spell checker but it doesn't catch this one

[the remark this answers is not in the document]
Comment (epage) [reply posts/inside-rust/2024-12-13-this-development-cycle-in-cargo-1.84.md:175]: https://github.com/crate-ci/typos/issues/1156#issuecomment-2539183786

File: posts/inside-rust/2024-12-13-this-development-cycle-in-cargo-1.84.md (hunk 4 above)
Comment (epage) [reply posts/inside-rust/2024-12-13-this-development-cycle-in-cargo-1.84.md:159]: We discussed it on the 3rd and merged on the 6th while this post only covers up to 2024-11-28.

Its less confusing when I get this posted right next to the release.  I'd like to get back to starting to right it before the release but this time there were the holidays.

**Review by ehuss — APPROVED** [review]
Thanks!

**Pull request merged**